README.md
# ping-exporter
[![Build Status](https://travis-ci.org/knsd/ping-exporter.svg?branch=master)](https://travis-ci.org/knsd/ping-exporter)

## Building and running

### Local Build

    make
    ./ping-exporter

Visiting [http://localhost:9346/ping?target=google.com](http://localhost:9346/ping?target=google.com) will return metrics for an ICMP ping against google.com.

### Building with Docker

    docker build -t ping-exporter .
    docker run -d -p 9346:9346 --name=ping-exporter ping-exporter

## Settings

| Environment variable                   | Default Value |
| -------------------------------------- | ------------- |
| PING_EXPORTER_LISTEN                   | [::]:9346     |
| PING_EXPORTER_DEFAULT_PROTOCOL         | v4            |
| PING_EXPORTER_RESOLVER                 | system        |
| PING_EXPORTER_DEFAULT_COUNT            | 5             |
| PING_EXPORTER_MAX_COUNT                | 30            |
| PING_EXPORTER_DEFAULT_PING_TIMEOUT     | 1000          |
| PING_EXPORTER_MAX_PING_TIMEOUT         | 10000         |
| PING_EXPORTER_DEFAULT_RESOLVE_TIMEOUT  | 1000          |
| PING_EXPORTER_MAX_RESOLVE_TIMEOUT      | 10000         |

## Available metrics

### `/ping` endpoint

| Metric name          | Type      | Description                                                                                    |
| -------------------- | --------- | ---------------------------------------------------------------------------------------------- |
| ping_resolve_error   | gauge     | Boolean metric if there's an error during the resolve (error message will be in "error" label) |
| ping_resolve_time    | gauge     | Time it take to resolve domain to an IP address                                                |
| ping_packets_total   | gauge     | Total number of sent pings                                                                     |
| ping_packets_success | gauge     | Total number of success pings                                                                  |
| ping_packets_failed  | gauge     | Total number of failed pings                                                                   |
| ping_packets_loss    | gauge     | A percentage of failed pings from the total pings                                              |
| ping_times           | histogram | A histogram of round-trip times                                                                |

### `/metrics` endpoint

| Metric name | Type    | Description                          |
| ----------- | ------- | ------------------------------------ |
| http_ping   | counter | Number of requests to /ping endpoint |

## Prometheus Configuration

This exporter needs to be passed the target as a parameter, this can be done with relabelling.

Example config:
```yml
scrape_configs:
  - job_name: 'ping'
    metrics_path: /ping
    static_configs:
      - targets:
        - google.com # Target to ping
    relabel_configs:
      - source_labels: [__address__]
        target_label: __param_target
      - source_labels: [__param_target]
        target_label: instance
      - target_label: __address__
        replacement: 127.0.0.1:9346 # This exporter's real hostname:port
```

In addition to that you can scrape the /metrics endpoint to be able to monitor exporter's own statistics.

Example config:
```yml
scrape_configs:
  - job_name: 'ping_exporter'
    static_configs:
      - targets:
        - 127.0.0.1:9346 # This exporter's real hostname:port
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## 0.3.0 - 2019-08-12

- Add `PING_EXPORTER_RESOLVER` option


## 0.2.0 - 2018-11-12

- Add `ping_resolve_error` metric


## 0.1.0 - 2018-06-15

- Initial release


src/http.rs
use std::net::IpAddr;

use futures::{future, Future};
use hyper::service::{NewService, Service};
use hyper::{Body, Method, Request, Response, Server, StatusCode};
use serde_urlencoded;
use tacho;

use metrics::{METRICS, REPORTER};
use pinger::{Pinger, Report};
use settings::Settings;
use utils::{boxed, NameOrIpAddr, Protocol};

lazy_static! {
    static ref HTTP_PING: tacho::Counter = METRICS.counter("http_ping", "Number of /ping requests");
}

pub fn init() {
    ::lazy_static::initialize(&HTTP_PING);
}

struct NewApp {
    settings: Settings,
    pinger: Pinger,
}

impl NewService for NewApp {
    type ReqBody = Body;
    type ResBody = Body;
    type Error = Box<::std::error::Error + Send + Sync>;
    type Service = App;
    type Future = future::FutureResult<Self::Service, Self::InitError>;
    type InitError = Box<::std::error::Error + Send + Sync>;

    fn new_service(&self) -> <Self as NewService>::Future {
        future::ok(App {
            settings: self.settings.clone(),
            pinger: self.pinger.clone(),
        })
    }
}

enum RequestType {
    Ping,
    Metrics,
    Unknown,
}

#[derive(Debug, Deserialize)]
struct PingRequest {
    target: NameOrIpAddr,
    protocol: Option<Protocol>,
    count: Option<usize>,
    ping_timeout: Option<u64>,
    resolve_timeout: Option<u64>,
}

struct App {
    settings: Settings,
    pinger: Pinger,
}

impl Service for App {
    type ReqBody = Body;
    type ResBody = Body;
    type Error = Box<::std::error::Error + Send + Sync>;
    type Future = Box<Future<Item = Response<Self::ResBody>, Error = Self::Error> + Send>;

    fn call(&mut self, req: Request<<Self as Service>::ReqBody>) -> <Self as Service>::Future {
        let st = ::time::precise_time_ns();
        let request_type = {
            let method = req.method();
            let path = req.uri().path();

            if method == &Method::GET && (path == "/ping" || path == "/ping/") {
                HTTP_PING.incr(1);
                RequestType::Ping
            } else if method == &Method::GET && (path == "/metrics" || path == "/metrics/") {
                RequestType::Metrics
            } else {
                RequestType::Unknown
            }
        };

        let future = match request_type {
            RequestType::Unknown => boxed(future::err((
                StatusCode::NOT_FOUND,
                Body::from("Not Found"),
            ))),
            RequestType::Metrics => boxed(get_metrics()),
            RequestType::Ping => {
                let query = req.uri().query().unwrap_or("");

                let mb_req = serde_urlencoded::from_str::<PingRequest>(query).map_err(|err| {
                    (
                        StatusCode::BAD_REQUEST,
                        Body::from(format!("Bad Request: {}", err)),
                    )
                });

                let future = future::result(mb_req);

                let settings = self.settings.clone();
                let pinger = self.pinger.clone();
                let future = future.and_then(move |request| ping(request, settings, pinger));
                boxed(future)
            }
        };

        let future = future.then(|request| match request {
            Err((status_code, body)) => {
                let mut response = Response::new(body);
                *response.status_mut() = status_code;
                future::ok(response)
            }
            Ok(body) => future::ok(Response::new(body)),
        });

        boxed(future.then(move |resp| {
            let delta = ::time::precise_time_ns() - st;
            let delta_ms = delta / 1000000;
            if let Some(path_and_query) = req.uri().path_and_query() {
                info!("{} {} {} ms", req.method(), path_and_query, delta_ms);
            } else {
                info!("{} {} {} ms", req.method(), req.uri().path(), delta_ms);
            }
            resp
        }))
    }
}

fn get_metrics() -> impl Future<Item = Body, Error = ((StatusCode, Body))> {
    future::result(format_metrics(&REPORTER.peek()))
}

fn ping(
    request: PingRequest,
    settings: Settings,
    pinger: Pinger,
) -> impl Future<Item = Body, Error = ((StatusCode, Body))> {
    let count = request.count.unwrap_or(settings.count);
    let ping_timeout = request.ping_timeout.unwrap_or(settings.ping_timeout);
    let resolve_timeout = request.resolve_timeout.unwrap_or(settings.resolve_timeout);

    let bad_request_body = {
        if count > settings.max_count {
            Some("Too many pings")
        } else if count < 1 {
            Some("Too few pings")
        } else if ping_timeout > settings.max_ping_timeout {
            Some("Too large ping timeout")
        } else if ping_timeout < 5 {
            Some("Too small ping timeout")
        } else if resolve_timeout > settings.resolve_timeout {
            Some("Too large resolve timeout")
        } else if resolve_timeout < 5 {
            Some("Too small resolve timeout")
        } else {
            None
        }
    };

    if let Some(body) = bad_request_body {
        return boxed(future::err((StatusCode::BAD_REQUEST, Body::from(body))));
    }

    let mut protocol = request.protocol.unwrap_or(settings.protocol);
    match &request.target {
        &NameOrIpAddr::IpAddr(IpAddr::V4(_)) => protocol = Protocol::V4,
        &NameOrIpAddr::IpAddr(IpAddr::V6(_)) => protocol = Protocol::V6,
        _ => (),
    }

    let name = request.target;

    let future = pinger.ping(name.clone(), protocol, count, resolve_timeout, ping_timeout);
    let future = future.map_err(|_| {
        let body = Body::from("Internal error");
        (StatusCode::INTERNAL_SERVER_ERROR, body)
    });

    let future = future.and_then(move |report| {
        let (metrics, reporter) = tacho::new();
        let metrics = metrics
            .labeled("target", name)
            .labeled("protocol", protocol)
            .labeled("count", count)
            .labeled("ping_timeout", ping_timeout)
            .labeled("resolve_timeout", resolve_timeout);

        set_metrics(metrics, report);

        format_metrics(&reporter.peek())
    });

    boxed(future)
}

pub fn set_metrics(mut metrics: tacho::Scope, report: Report) {
    let resolve_error: Option<&'static str>;

    match report {
        Report::ResolveTimedOut => resolve_error = Some("timed out"),
        Report::ResolveNotFound => resolve_error = Some("not found"),
        Report::ResolveOtherError => resolve_error = Some("internal error"),
        Report::Success {
            resolve_time_ns,
            addr,
            pings,
        } => {
            resolve_error = None;
            metrics = metrics.labeled("ip", addr);

            metrics
                .gauge(
                    "ping_resolve_time",
                    "Time it take to resolve domain to an IP address",
                )
                .set((resolve_time_ns / 1000000) as usize);

            let times = metrics.stat("ping_times", "A histogram of round-trip times");

            let mut failures = 0;
            let mut successful = 0;
            let total = pings.len();

            for reply_time in pings {
                match reply_time {
                    Some(reply_time) => {
                        times.add((reply_time * 1000.0) as u64);
                        successful += 1;
                    }
                    None => {
                        failures += 1;
                    }
                }
            }

            metrics
                .gauge("ping_packets_total", "Total number of sent pings")
                .set(total);
            metrics
                .gauge("ping_packets_success", "Total number of success pings")
                .set(successful);
            metrics
                .gauge("ping_packets_failed", "Total number of failed pings")
                .set(failures);

            if total > 0 {
                let loss = failures as f64 / total as f64 * 100.0;
                metrics
                    .gauge(
                        "ping_packets_loss",
                        "A percentage of failed pings from the total pings",
                    )
                    .set(loss as usize);
            }
        }
    };

    if let Some(error) = resolve_error {
        metrics = metrics.labeled("error", error);
    }

    metrics.gauge("ping_resolve_error", "Boolean metric if there's an error during the resolve (error message will be in \"error\" label)")
        .set(resolve_error.map(|_| 1).unwrap_or(0));
}

fn format_metrics(report: &tacho::Report) -> Result<Body, (StatusCode, Body)> {
    match tacho::prometheus::string(report) {
        Ok(s) => Ok(Body::from(s)),
        Err(_) => Err((
            StatusCode::INTERNAL_SERVER_ERROR,
            Body::from("Internal Error"),
        )),
    }
}

pub fn server(settings: Settings, pinger: Pinger) -> impl Future<Item = (), Error = ()> {
    let builder = Server::try_bind(&settings.listen);
    let future = future::result(builder).and_then(move |builder| {
        info!("Listening on {}", &settings.listen);
        builder.serve(NewApp { settings, pinger })
    });
    let future = future.map_err(|error| {
        error!("Server error: {}", error);
    });
    future
}

#[cfg(test)]
mod tests {
    use super::init;

    #[test]
    fn test_lazy_static() {
        init()
    }
}


src/main.rs
#![recursion_limit = "128"]

#[macro_use]
extern crate failure;
extern crate futures;
extern crate hyper;
#[macro_use]
extern crate lazy_static;
extern crate rand;
extern crate serde;
#[macro_use]
extern crate serde_derive;
extern crate serde_urlencoded;
#[macro_use]
extern crate slog;
#[macro_use]
extern crate log;
extern crate slog_async;
extern crate slog_scope;
extern crate slog_stdlog;
extern crate slog_term;
extern crate tacho;
extern crate time;
extern crate tokio;
extern crate tokio_ping;
extern crate tokio_signal;
extern crate trust_dns_resolver;

use futures::future::Either;
use futures::sync::oneshot;
use futures::{Future, Stream};
use slog::Drain;
use tokio_signal::unix::{Signal, SIGINT, SIGTERM};

mod http;
mod metrics;
mod pinger;
mod resolver;
mod settings;
mod utils;

fn signals() -> impl Future<Item = i32, Error = ::std::io::Error> {
    futures::future::select_all((&[SIGTERM, SIGINT]).iter().map(|&signum| {
        Signal::new(signum)
            .flatten_stream()
            .into_future()
            .map(|(signum, _rest)| signum)
            .map_err(|(err, _rest)| err)
    }))
    .then(|res| match res {
        Ok((Some(signum), _, _)) => Ok(signum),
        Err((err, _, _)) => Err(err),
        Ok((None, _, _)) => unreachable!("signals"),
    })
}

fn run() -> i32 {
    let decorator = slog_term::TermDecorator::new().build();
    let drain = slog_term::FullFormat::new(decorator).build().fuse();
    let drain = slog_async::Async::new(drain).chan_size(4096).build().fuse();
    let drain = slog::LevelFilter::new(drain, slog::Level::Info).fuse();
    let log = slog::Logger::root(drain, o!());
    let _scope_guard = slog_scope::set_global_logger(log.new(o!()));
    slog_stdlog::init().expect("Init std logger");

    info!(concat!(
        "Starting ",
        env!("CARGO_PKG_NAME"),
        " ",
        env!("CARGO_PKG_VERSION")
    ));

    let settings = match settings::Settings::from_env() {
        Ok(settings) => settings,
        Err(err) => {
            error!("{}", err);
            return 1;
        }
    };

    info!("Using settings: {}", settings);

    http::init();
    metrics::init();

    let mut runtime = tokio::runtime::Runtime::new().expect("Tokio runtime");
    let (stop_sender, stop_receiver) = oneshot::channel();

    runtime.spawn(futures::lazy(move || {
        let server_future = pinger::Pinger::new(settings.clone())
            .map_err(|_| {
                error!("Unable to create pinger, please check capabilities");
            })
            .and_then(move |pinger| http::server(settings, pinger));

        let signals_future = signals().map_err(|_| {
            error!("Signal handling error");
        });

        signals_future.select2(server_future).then(move |res| {
            if let Ok(Either::A((signum, _))) = res {
                info!("Got signal {}.", signum)
            }
            stop_sender.send(()).ok();
            Ok(())
        })
    }));

    stop_receiver.wait().ok();
    runtime.shutdown_now().wait().ok();
    info!("Exiting");
    0
}

fn main() {
    ::std::process::exit(run())
}


src/utils.rs
use std::fmt;
use std::net::IpAddr;
use std::str::FromStr;
use std::sync::Arc;

use futures::Future;
use serde::{de::Error as SerdeDeError, Deserialize, Deserializer};
use trust_dns_resolver::Name;

#[derive(Debug, Clone, Copy)]
pub enum Protocol {
    V4,
    V6,
}

impl fmt::Display for Protocol {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            &Protocol::V4 => write!(f, "v4"),
            &Protocol::V6 => write!(f, "v6"),
        }
    }
}

impl FromStr for Protocol {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, <Self as FromStr>::Err> {
        match s {
            "v4" => Ok(Protocol::V4),
            "v6" => Ok(Protocol::V6),
            other => Err(format!("'{}' is not valid protocol, use v4 or v6", other)),
        }
    }
}

impl<'de> Deserialize<'de> for Protocol {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Protocol::from_str(String::deserialize(deserializer)?.as_str()).map_err(D::Error::custom)
    }
}

#[derive(Debug, Clone)]
pub enum NameOrIpAddr {
    Name(Arc<Name>),
    IpAddr(IpAddr),
}

impl fmt::Display for NameOrIpAddr {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            &NameOrIpAddr::Name(ref name) => name.fmt(f),
            &NameOrIpAddr::IpAddr(ref addr) => addr.fmt(f),
        }
    }
}

impl FromStr for NameOrIpAddr {
    type Err = <Name as FromStr>::Err;

    fn from_str(s: &str) -> Result<Self, <Self as FromStr>::Err> {
        IpAddr::from_str(s)
            .map(|addr| NameOrIpAddr::IpAddr(addr))
            .or_else(|_| Ok(NameOrIpAddr::Name(Arc::new(Name::from_str(s)?))))
    }
}

impl<'de> Deserialize<'de> for NameOrIpAddr {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        NameOrIpAddr::from_str(String::deserialize(deserializer)?.as_str())
            .map_err(|_| D::Error::custom("Invalid hostname or IP addr"))
    }
}

pub fn boxed<F: Future<Item = I, Error = E> + Send + 'static, I, E>(
    future: F,
) -> Box<Future<Item = I, Error = E> + Send> {
    Box::new(future)
}


src/pinger.rs
use std::net::IpAddr;
use std::sync::Arc;
use std::time::Duration;

use futures::future::{ok, Either};
use futures::{future, Future, Stream};
use tokio::timer::Timeout;
use tokio_ping::{Error as PingError, Pinger as LowLevelPinger};

use resolver::{Error as ResolveError, Resolver};
use settings::Settings;
use utils::{NameOrIpAddr, Protocol};

#[derive(Debug, Fail)]
pub enum Error {
    #[fail(display = "ping error")]
    PingError { error: PingError },
    #[fail(display = "create resolver error")]
    CreateResolverError { error: ResolveError },
}

impl From<PingError> for Error {
    fn from(error: PingError) -> Self {
        Error::PingError { error }
    }
}

impl From<ResolveError> for Error {
    fn from(error: ResolveError) -> Self {
        Error::CreateResolverError { error }
    }
}

pub enum Report {
    ResolveNotFound,
    ResolveTimedOut,
    ResolveOtherError,
    Success {
        resolve_time_ns: u64,
        addr: IpAddr,
        pings: Vec<Option<f64>>,
    },
}

#[derive(Clone)]
pub struct Pinger {
    inner: Arc<PingerInner>,
}

struct PingerInner {
    resolver: Resolver,
    pinger: LowLevelPinger,
}

impl Pinger {
    pub fn new(settings: Settings) -> impl Future<Item = Self, Error = Error> {
        let resolver_future = Resolver::new(settings).map_err(From::from);
        let pinger_future = LowLevelPinger::new().map_err(From::from);
        resolver_future
            .join(pinger_future)
            .and_then(|(resolver, pinger)| {
                Ok(Self {
                    inner: Arc::new(PingerInner { resolver, pinger }),
                })
            })
    }

    pub fn ping(
        &self,
        name: NameOrIpAddr,
        protocol: Protocol,
        count: usize,
        resolve_timeout: u64,
        timeout: u64,
    ) -> impl Future<Item = Report, Error = Error> {
        let resolve_timeout = Duration::from_millis(resolve_timeout);
        let timeout = Duration::from_millis(timeout);

        let future = self.inner.resolver.resolve(name, protocol);
        let future = Timeout::new(future, resolve_timeout);

        let pinger = self.inner.pinger.clone();
        let future = future.then(move |result| match result {
            Ok((resolve_time_ns, addr)) => {
                let stream = pinger
                    .chain(addr)
                    .timeout(timeout)
                    .stream()
                    .map_err(From::from)
                    .take(count as u64);

                let future = stream.fold(Vec::new(), |mut acc, result| {
                    acc.push(result);
                    future::ok::<Vec<Option<f64>>, Error>(acc)
                });

                Either::A(future.and_then(move |pings| {
                    Ok(Report::Success {
                        resolve_time_ns,
                        addr,
                        pings,
                    })
                }))
            }
            Err(err) => Either::B(ok(match err.into_inner() {
                Some(ResolveError::NotFound) => Report::ResolveNotFound,
                Some(ResolveError::Error) => Report::ResolveOtherError,
                None => Report::ResolveTimedOut,
            })),
        });

        future.and_then(|report| Ok(report))
        //
        //        let future = future.map_err(|err| match err.into_inner() {
        //            Some(err) => err.into(),
        //            None => Error::ResolveTimeoutError {}
        //        });
        //
        //        let pinger = self.inner.pinger.clone();
        //        let future = future.and_then(move |(resolve_time_ns, addr)| {
        //            let stream = pinger.chain(addr)
        //                .timeout(timeout)
        //                .stream()
        //                .map_err(From::from)
        //                .take(count as u64);
        //
        //            let future = stream.fold(Vec::new(), |mut acc, result| {
        //                acc.push(result);
        //                future::ok::<Vec<Option<f64>>, Error>(acc)
        //            });
        //
        //            future.and_then(move |times| {
        //                Ok(Report::Success {
        //                    resolve_time_ns,
        //                    addr,
        //                    times,
        //                })
        //            })
        //        });
    }
}


src/logging.rs
use slog;
use slog::Drain;

use std::sync::atomic;
use std::sync::atomic::Ordering;
use std::result;

pub static DEBUG: atomic::AtomicBool = atomic::ATOMIC_BOOL_INIT;

struct RuntimeLevelFilter<D>{
    drain: D,
}

impl<D> Drain for RuntimeLevelFilter<D>
    where D : Drain {
    type Ok = Option<D::Ok>;
    type Err = Option<D::Err>;

    fn log(&self,
           record: &slog::Record,
           values: &slog::OwnedKVList)
           -> result::Result<Self::Ok, Self::Err> {
        let current_level = if DEBUG.load(Ordering::Relaxed) {
            slog::Level::Debug
        } else {
            slog::Level::Info
        };

        if record.level().is_at_least(current_level) {
            self.drain.log(
                record,
                values
            )
                .map(Some)
                .map_err(Some)
        } else {
            Ok(None)
        }
    }
}

pub fn init() -> slog::Logger {
    let decorator = ::slog_term::PlainDecorator::new(::std::io::stdout());
    let drain = ::slog_term::CompactFormat::new(decorator).build().fuse();
    let drain = RuntimeLevelFilter { drain: drain }.fuse();
    let drain = ::slog_async::Async::new(drain).chan_size(4096).build().fuse();
    slog::Logger::root(drain, o!())
}


src/resolver.rs
use std::net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr};

use futures::{future, Future};
use rand::{seq::SliceRandom, thread_rng};
use trust_dns_resolver::config::{self, NameServerConfig, ResolverConfig, ResolverOpts};
use trust_dns_resolver::error::{ResolveError, ResolveErrorKind};
use trust_dns_resolver::AsyncResolver;

use settings::Settings;
use utils::{boxed, NameOrIpAddr, Protocol};

pub struct Resolver {
    inner: AsyncResolver,
}

#[derive(Debug, Fail)]
pub enum Error {
    #[fail(display = "not found")]
    NotFound,
    #[fail(display = "unknown error")]
    Error,
}

impl From<ResolveError> for Error {
    fn from(err: ResolveError) -> Self {
        if let ResolveErrorKind::NoRecordsFound { .. } = err.kind() {
            Error::NotFound
        } else {
            Error::Error
        }
    }
}

impl Resolver {
    pub fn new(settings: Settings) -> impl Future<Item = Self, Error = Error> {
        let future = future::result(match settings.resolver {
            Some(resolver_addr) => {
                let mut config = ResolverConfig::new();
                config.add_name_server(NameServerConfig {
                    socket_addr: SocketAddr::new(resolver_addr, 53),
                    protocol: config::Protocol::Udp,
                    tls_dns_name: None,
                });
                let (client, future) = AsyncResolver::new(config, ResolverOpts::default());
                tokio::spawn(future);
                Ok(client)
            }
            None => AsyncResolver::from_system_conf().map(|(client, future)| {
                tokio::spawn(future);
                client
            }),
        });

        future
            .map_err(From::from)
            .and_then(|inner| Ok(Resolver { inner }))
    }

    pub fn resolve(
        &self,
        name: NameOrIpAddr,
        protocol: Protocol,
    ) -> impl Future<Item = (u64, IpAddr), Error = Error> {
        let st = ::time::precise_time_ns();
        match name {
            NameOrIpAddr::IpAddr(addr) => boxed(future::ok((0, addr))),
            NameOrIpAddr::Name(name) => {
                let future = match protocol {
                    Protocol::V4 => boxed(
                        self.inner
                            .ipv4_lookup(name.as_ref().clone())
                            .map_err(From::from)
                            .and_then(|addrs| {
                                let addrs: Vec<Ipv4Addr> = addrs.iter().cloned().collect();

                                let random_addr =
                                    addrs.choose(&mut thread_rng()).ok_or(Error::NotFound)?;
                                Ok(IpAddr::from(random_addr.clone()))
                            }),
                    ),
                    Protocol::V6 => boxed(
                        self.inner
                            .ipv6_lookup(name.as_ref().clone())
                            .map_err(From::from)
                            .and_then(|addrs| {
                                let addrs: Vec<Ipv6Addr> = addrs.iter().cloned().collect();
                                let random_addr =
                                    addrs.choose(&mut thread_rng()).ok_or(Error::NotFound)?;

                                Ok(IpAddr::from(random_addr.clone()))
                            }),
                    ),
                };

                boxed(future.and_then(move |addr| Ok((::time::precise_time_ns() - st, addr))))
            }
        }
    }
}


src/settings.rs
use std::env;
use std::fmt;
use std::net::{IpAddr, SocketAddr};
use std::ops::Deref;
use std::str::FromStr;
use std::sync::Arc;

use utils::Protocol;

static ENV_PREFIX: &str = "PING_EXPORTER";
static ENV_SEPARATOR: &str = "_";

lazy_static! {
    static ref DEFAULT_LISTEN: SocketAddr =
        SocketAddr::from_str("[::]:9346").expect("DEFAULT_LISTEN");
}

#[derive(Debug, Clone)]
pub struct Settings {
    inner: Arc<SettingsInner>,
}

impl fmt::Display for Settings {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        write!(f, "listen address: {}, ", self.listen)?;
        write!(f, "preferred protocol: {}, ", self.protocol)?;
        match self.resolver {
            Some(resolver) => write!(f, "resolver: {}, ", resolver)?,
            None => write!(f, "resolver: system, ")?,
        }
        write!(f, "default number of ICMP packets: {}, ", self.count)?;
        write!(f, "maximum number of ICMP packets: {}, ", self.max_count)?;
        write!(
            f,
            "timeout for each ICMP packet: {} ms, ",
            self.ping_timeout
        )?;
        write!(
            f,
            "maximum timeout for each ICMP packet: {} ms, ",
            self.max_ping_timeout
        )?;
        write!(f, "resolve timeout: {} ms, ", self.resolve_timeout)?;
        write!(
            f,
            "maximum resolve timeout: {} ms.",
            self.max_resolve_timeout
        )?;
        Ok(())
    }
}

#[derive(Debug)]
pub struct SettingsInner {
    pub listen: SocketAddr,
    pub protocol: Protocol,
    pub resolver: Option<IpAddr>,
    pub count: usize,
    pub max_count: usize,
    pub ping_timeout: u64,
    pub max_ping_timeout: u64,
    pub resolve_timeout: u64,
    pub max_resolve_timeout: u64,
}

impl Settings {
    pub fn from_env() -> Result<Self, Error> {
        Ok(Self {
            inner: Arc::new(SettingsInner {
                listen: get_env_or("LISTEN", DEFAULT_LISTEN.clone())?,
                protocol: get_env_or("DEFAULT_PROTOCOL", Protocol::V4)?,
                resolver: match get_env_("RESOLVER") {
                    Ok(resolver) => Some(resolver),
                    Err(Error::MissingEnvVar { .. }) => None,
                    Err(err) => return Err(err.into()),
                },
                count: get_env_or("DEFAULT_COUNT", 5)?,
                max_count: get_env_or("MAX_COUNT", 30)?,
                ping_timeout: get_env_or("DEFAULT_PING_TIMEOUT", 1000)?,
                max_ping_timeout: get_env_or("MAX_PING_TIMEOUT", 10000)?,
                resolve_timeout: get_env_or("DEFAULT_RESOLVE_TIMEOUT", 1000)?,
                max_resolve_timeout: get_env_or("MAX_RESOLVE_TIMEOUT", 10000)?,
            }),
        })
    }
}

impl Deref for Settings {
    type Target = SettingsInner;

    fn deref(&self) -> &<Self as Deref>::Target {
        &self.inner
    }
}

#[derive(Debug, Fail)]
pub enum Error {
    #[fail(display = "missing environment variable: {}", name)]
    MissingEnvVar { name: String },
    #[fail(display = "invalid environment variable type: {}", name)]
    InvalidVariableType { name: String },
}

fn get_env_or<T: FromStr>(name: &str, default: T) -> Result<T, Error> {
    match get_env_(name) {
        Ok(v) => Ok(v),
        Err(Error::MissingEnvVar { .. }) => Ok(default),
        Err(err) => Err(err),
    }
}

fn get_env_<T: FromStr>(name: &str) -> Result<T, Error> {
    let env_var_name = format!("{}{}{}", ENV_PREFIX, ENV_SEPARATOR, name.to_uppercase());

    let string = env::var(&env_var_name).map_err(|_| Error::MissingEnvVar {
        name: env_var_name.clone(),
    })?;
    T::from_str(&string).map_err(|_| Error::InvalidVariableType { name: env_var_name })
}

#[cfg(test)]
mod tests {
    use super::Settings;

    #[test]
    fn test_valid_default_settings() {
        assert!(Settings::from_env().is_ok());
    }
}
